app/models/ems_refresh/save_inventory_infra.rb

```diff
@@ -44,6 +44,19 @@ def save_ems_infra_inventory(ems, hashes, target = nil)
       return
     end
 
+    # Check if a vm was removed and we still have disk(s)
+    if target.kind_of?(VmOrTemplate) && hashes[:vms].blank?
+      current = hashes[:vm_disks].collect { |disk| URI(disk[:storage][:ems_ref]).path.split('/').last }.uniq
+      if current.blank?
+        storage = nil
+      else
+        vm_storages = ([target.storage] + target.storages).compact.uniq
+        storage = vm_storages.select { |store| !current.include?(store.ems_ref) }
+      end
+
+      target.disconnect_storage storage
+    end
```

Comment:
This feels wrong here in the save_inventory.  save_inventory should be mostly data-driven and provider agnostic.  vm_disks is not a normal top-level key for other providers.  I would expect the parser to deal with this by not attaching the storage in the data it provides.